storage/Devices/MsdosImpl.cc

```diff
@@ -63,6 +63,29 @@ namespace storage
 
 
     void
+    Msdos::Impl::delete_partition(Partition* partition)
+    {
+	PartitionType old_type = partition->get_type();
+	unsigned int old_number = partition->get_number();
+
+	PartitionTable::Impl::delete_partition(partition);
+
+	// After deleting a logical partition the numbers of partitions with
+	// higher numbers are shifted.
+	if (old_type == PartitionType::LOGICAL)
+	{
+	    vector<Partition*> partitions = get_partitions();
+            for (Partition* tmp : partitions)
+	    {
+		unsigned int number = tmp->get_number();
+		if (number > old_number)
+		    tmp->get_impl().set_number(number - 1);
```

Comment:
Similarly, I wouldn't rely on that partition's old number; I'd keep my own counter.